File: libreria/EJEMPLOS/Benchmarks/benchmark_latencia.py
# NEVEN Benchmark — Latencia de Llamadas Python
# Ejecutar desde Excel: =NEVEN.P("exec(open('benchmark_latencia.py').read())")
# O usar las funciones individuales via =P.func()

import time
import random
import math

def NEVEN_Benchmark_Basico_Py(n_iteraciones=100):
    """Benchmark: mide latencia basica de Python (N iteraciones de 1+1).
    
    Args:
        n_iteraciones: Numero de iteraciones (default: 100)
    """
    tiempos = []
    for _ in range(int(n_iteraciones)):
        t0 = time.perf_counter()
        resultado = 1 + 1
        t1 = time.perf_counter()
        tiempos.append((t1 - t0) * 1000)  # milisegundos
    
    tiempos.sort()
    n = len(tiempos)
    media = sum(tiempos) / n
    mediana = tiempos[n // 2]
    p95 = tiempos[int(n * 0.95)]
    p99 = tiempos[int(n * 0.99)]
    
    return [
        ["Metrica", "Valor"],
        ["N", n],
        ["Media_ms", round(media, 6)],
        ["Mediana_ms", round(mediana, 6)],
        ["Min_ms", round(min(tiempos), 6)],
        ["Max_ms", round(max(tiempos), 6)],
        ["P95_ms", round(p95, 6)],
        ["P99_ms", round(p99, 6)]
    ]


def NEVEN_Benchmark_Regresion_Py(n_filas=1000):
    """Benchmark: regresion lineal simple en Python puro (sin numpy).
    
    Args:
        n_filas: Numero de filas de datos (default: 1000)
    """
    random.seed(42)
    n = int(n_filas)
    x = [random.gauss(0, 1) for _ in range(n)]
    y = [2 * xi + random.gauss(0, 0.5) for xi in x]
    
    t0 = time.perf_counter()
    
    # Regresion lineal por minimos cuadrados
    sum_x = sum(x)
    sum_y = sum(y)
    sum_xy = sum(xi * yi for xi, yi in zip(x, y))
    sum_x2 = sum(xi ** 2 for xi in x)
    
    pendiente = (n * sum_xy - sum_x * sum_y) / (n * sum_x2 - sum_x ** 2)
    intercepto = (sum_y - pendiente * sum_x) / n
    
    # R-cuadrado
    y_pred = [intercepto + pendiente * xi for xi in x]
    ss_res = sum((yi - yp) ** 2 for yi, yp in zip(y, y_pred))
    y_mean = sum_y / n
    ss_tot = sum((yi - y_mean) ** 2 for yi in y)
    r2 = 1 - ss_res / ss_tot
    
    t1 = time.perf_counter()
    tiempo_ms = (t1 - t0) * 1000
    
    return [
        ["Metrica", "Valor"],
        ["N_filas", n],
        ["Tiempo_ms", round(tiempo_ms, 2)],
        ["R2", round(r2, 6)],
        ["Intercepto", round(intercepto, 6)],
        ["Pendiente", round(pendiente, 6)]
    ]


def NEVEN_Benchmark_Escalabilidad_Py():
    """Benchmark: escalabilidad de regresion lineal Python (100 a 100K filas)."""
    tamanos = [100, 500, 1000, 5000, 10000, 50000, 100000]
    resultados = [["N_filas", "Tiempo_ms"]]
    
    for n in tamanos:
        random.seed(42)
        x = [random.gauss(0, 1) for _ in range(n)]
        y = [2 * xi + random.gauss(0, 0.5) for xi in x]
        
        t0 = time.perf_counter()
        
        sum_x = sum(x)
        sum_y = sum(y)
        sum_xy = sum(xi * yi for xi, yi in zip(x, y))
        sum_x2 = sum(xi ** 2 for xi in x)
        pendiente = (n * sum_xy - sum_x * sum_y) / (n * sum_x2 - sum_x ** 2)
        intercepto = (sum_y - pendiente * sum_x) / n
        
        t1 = time.perf_counter()
        resultados.append([n, round((t1 - t0) * 1000, 2)])
    
    return resultados

File: libreria/EJEMPLOS/Benchmarks/benchmark_latencia.R
# NEVEN Benchmark — Latencia de Llamadas
# Ejecutar desde Excel: =NEVEN.R("source('benchmark_latencia.R')")
# O copiar y pegar cada bloque en celdas individuales.
#
# Este script mide el tiempo de ida y vuelta (round-trip) de llamadas
# desde Excel a R via Named Pipes + Protobuf.

# ─── 1. Latencia basica (operacion trivial) ─────────────────────────

NEVEN_Benchmark_Basico <- function(n_iteraciones = 100) {
  tiempos <- numeric(n_iteraciones)
  for (i in 1:n_iteraciones) {
    t0 <- proc.time()["elapsed"]
    resultado <- 1 + 1  # Operacion trivial
    t1 <- proc.time()["elapsed"]
    tiempos[i] <- (t1 - t0) * 1000  # milisegundos
  }
  
  data.frame(
    Metrica = c("N", "Media_ms", "Mediana_ms", "Min_ms", "Max_ms", "P95_ms", "P99_ms"),
    Valor = c(
      n_iteraciones,
      round(mean(tiempos), 4),
      round(median(tiempos), 4),
      round(min(tiempos), 4),
      round(max(tiempos), 4),
      round(quantile(tiempos, 0.95), 4),
      round(quantile(tiempos, 0.99), 4)
    )
  )
}

# ─── 2. Latencia con datos (regresion lineal) ───────────────────────

NEVEN_Benchmark_Regresion <- function(n_filas = 1000) {
  set.seed(42)
  x <- rnorm(n_filas)
  y <- 2 * x + rnorm(n_filas, sd = 0.5)
  
  t0 <- proc.time()["elapsed"]
  modelo <- lm(y ~ x)
  coefs <- coef(modelo)
  r2 <- summary(modelo)$r.squared
  t1 <- proc.time()["elapsed"]
  
  tiempo_ms <- (t1 - t0) * 1000
  
  data.frame(
    Metrica = c("N_filas", "Tiempo_ms", "R2", "Intercepto", "Pendiente"),
    Valor = c(n_filas, round(tiempo_ms, 2), round(r2, 6), round(coefs[1], 6), round(coefs[2], 6))
  )
}

# ─── 3. Benchmark de escalabilidad ──────────────────────────────────

NEVEN_Benchmark_Escalabilidad <- function() {
  tamanos <- c(100, 500, 1000, 5000, 10000, 50000, 100000)
  resultados <- data.frame(N_filas = integer(), Tiempo_ms = numeric())
  
  for (n in tamanos) {
    set.seed(42)
    x <- rnorm(n)
    y <- 2 * x + rnorm(n, sd = 0.5)
    
    t0 <- proc.time()["elapsed"]
    modelo <- lm(y ~ x)
    t1 <- proc.time()["elapsed"]
    
    resultados <- rbind(resultados, data.frame(
      N_filas = n,
      Tiempo_ms = round((t1 - t0) * 1000, 2)
    ))
  }
  
  resultados
}

attr(NEVEN_Benchmark_Basico, "description") <- list(
  "Benchmark: mide latencia basica de R (N iteraciones de 1+1)",
  n_iteraciones = "Numero de iteraciones (default: 100)"
)
attr(NEVEN_Benchmark_Basico, "category") <- "NEVEN Benchmarks"

attr(NEVEN_Benchmark_Regresion, "description") <- list(
  "Benchmark: mide tiempo de regresion lineal con N filas",
  n_filas = "Numero de filas de datos (default: 1000)"
)
attr(NEVEN_Benchmark_Regresion, "category") <- "NEVEN Benchmarks"

attr(NEVEN_Benchmark_Escalabilidad, "description") <- list(
  "Benchmark: mide escalabilidad de regresion lineal (100 a 100K filas)"
)
attr(NEVEN_Benchmark_Escalabilidad, "category") <- "NEVEN Benchmarks"
